File: ts/src/cli/transcribe.ts
#!/usr/bin/env ts-node
import { Command } from "commander";
import fs from "node:fs";
import path from "node:path";
import { DEFAULT_BASE_URL, DEFAULT_POLL_INTERVAL } from "../api";
import { transcribeToFile } from "../transcriber";

function buildCommand(): Command {
  const program = new Command();
  program
    .description("Submit audio to Soniox and save the JSON transcript.")
    .option(
      "--audio <path>",
      "Path to the local audio file to transcribe",
      "audio.wav"
    )
    .option(
      "--audio-url <url>",
      "Optional public URL of the audio file instead of --audio."
    )
    .option(
      "--model <name>",
      "Soniox model to use",
      "stt-async-preview"
    )
    .option(
      "--output <path>",
      "Path to save the JSON response",
      "response.json"
    )
    .option("--keep-resources", "Skip deleting the uploaded file and transcription on Soniox.", false)
    .option(
      "--poll-interval <seconds>",
      "Seconds between status polls",
      (value) => parseFloat(value),
      DEFAULT_POLL_INTERVAL
    )
    .option("--base-url <url>", "Override the Soniox API base URL", DEFAULT_BASE_URL);
  return program;
}

async function main(argv: string[]): Promise<number> {
  const program = buildCommand();
  program.exitOverride();

  try {
    const options = program.parse(argv).opts<{
      audio: string;
      audioUrl?: string;
      model: string;
      output: string;
      keepResources: boolean;
      pollInterval: number;
      baseUrl: string;
    }>();

    if (!options.audioUrl && !options.audio) {
      throw new Error("Provide either --audio or --audio-url.");
    }
    if (!options.audioUrl) {
      const audioPath = path.resolve(options.audio);
      if (!fs.existsSync(audioPath)) {
        throw new Error(`Audio file not found: ${audioPath}`);
      }
      console.info(`Using local audio file ${audioPath}`);
    } else {
      console.info(`Using remote audio URL ${options.audioUrl}`);
    }

    await transcribeToFile({
      audioPath: options.audioUrl ? undefined : options.audio,
      audioUrl: options.audioUrl,
      model: options.model,
      outputPath: options.output,
      pollInterval: options.pollInterval,
      keepRemote: options.keepResources,
      baseUrl: options.baseUrl
    });
    console.info(`Saved transcription JSON to ${path.resolve(options.output)}`);
    return 0;
  } catch (error) {
    if (error instanceof Error && (error as any).code === "commander.helpDisplayed") {
      return 0;
    }
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }
}

if (require.main === module) {
  main(process.argv).then(
    (code) => {
      process.exit(code);
    },
    (err) => {
      console.error(err instanceof Error ? err.message : String(err));
      process.exit(1);
    }
  );
}

export default main;

File: sonioxsrt/cli/transcribe.py
"""CLI for submitting audio to Soniox and saving the JSON transcript."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Optional, Sequence

from ..api import DEFAULT_BASE_URL, DEFAULT_POLL_INTERVAL, SonioxError
from ..transcriber import transcribe_to_file


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Submit audio to Soniox and save the JSON transcript."
    )
    parser.add_argument(
        "--audio",
        default="audio.wav",
        help="Path to the local audio file to transcribe (default: audio.wav).",
    )
    parser.add_argument(
        "--audio-url",
        help="Optional public URL of the audio file instead of --audio.",
    )
    parser.add_argument(
        "--model",
        default="stt-async-preview",
        help="Soniox model to use (default: stt-async-preview).",
    )
    parser.add_argument(
        "--output",
        default="response.json",
        help="Path to save the JSON response (default: response.json).",
    )
    parser.add_argument(
        "--keep-resources",
        action="store_true",
        help="Skip deleting the uploaded file and transcription on Soniox.",
    )
    parser.add_argument(
        "--poll-interval",
        type=float,
        default=DEFAULT_POLL_INTERVAL,
        help=(
            f"Seconds between status polls (default: {DEFAULT_POLL_INTERVAL}). "
            "Lower values poll more frequently."
        ),
    )
    parser.add_argument(
        "--base-url",
        default=DEFAULT_BASE_URL,
        help=f"Override the Soniox API base URL (default: {DEFAULT_BASE_URL}).",
    )
    return parser


def main(argv: Optional[Sequence[str]] = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)

    if not args.audio_url and not Path(args.audio).exists():
        parser.error(f"Audio file not found: {args.audio}")

    try:
        transcribe_to_file(
            audio_path=None if args.audio_url else Path(args.audio),
            audio_url=args.audio_url,
            model=args.model,
            output_path=Path(args.output),
            poll_interval=args.poll_interval,
            keep_remote=args.keep_resources,
            base_url=args.base_url,
        )
    except (FileNotFoundError, ValueError) as exc:
        print(exc, file=sys.stderr)
        return 1
    except SonioxError as exc:
        print(exc, file=sys.stderr)
        return 1

    print(f"Saved transcription JSON to {args.output}")
    return 0


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main())
